feat: add ptp function to statistics module (#1897)

ivansherbakov9 opened the pull request:
## Due Diligence
- General:
    - [x] **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [ ] unit tests: MPS tested (1 MPI process, 1 GPU)
    - [ ] benchmarks: created for new functionality
    - [ ] benchmarks: performance improved or maintained
    - [x] documentation updated where needed

## Description

Implemented `ht.ptp`, the peak-to-peak function, consistent with `numpy.ptp`.
Supports `axis`, `out`, and `keepdims` arguments.

Closes #1897

Issue/s resolved: #1897

## Changes proposed:
- Added `ptp` function to `heat/core/statistics.py`
- Added `DNDarray.ptp` method binding
- Added unit tests for `ptp` in `heat/core/tests/test_statistics.py`
- Updated docstrings and examples

## Type of change
- New feature (non-breaking change which adds functionality)

## Memory requirements
No additional memory requirements compared to equivalent `ht.max` and `ht.min`.

## Performance
Performance is expected to be similar to running `ht.max` and `ht.min` separately.  
Future optimization may reduce collective communications to a single pass.

#### Does this change modify the behaviour of other functions? 
No.

ClaudiaComito commented:
> @JuanPedroGHM @ClaudiaComito Thanks, I've applied the suggested fix 👍

@ivansherbakov9 thanks a lot for your contribution! Do you want to try and write the custom reduction operation to minimize communication?

ivansherbakov9 commented:
@ClaudiaComito Thanks for the review!  

I'd like to try implementing the custom MPI reduction for ptp, to minimize communication as you suggested.  
I'll experiment with it and ask here (or in Mattermost) if I get stuck.

ivansherbakov9 commented:
@ClaudiaComito Thanks for the suggestions. I applied a subset of Copilot's proposals:
* Changed callback annotation to memoryview for clarity
* Narrowed exception handling to (ValueError, TypeError).
* Replaced the lambda binding with an explicit _ptp wrapper for readability

I intentionally kept the candidate dtype list explicit (extended to common types) rather than generating it dynamically because the dynamic approach was brittle and hard to read. If you prefer a different ordering of dtypes (e.g. int64 first) I can adjust.

JuanPedroGHM commented on heat/core/statistics.py:
```diff
@@ -1460,6 +1461,86 @@ def mpi_argmin(a: str, b: str, _: Any) -> torch.Tensor:
+        placeholder
+    """
+    # Candidate dtypes to try
+    candidates = (
+        np.float64,
+        np.float32,
+        np.float16,
+        np.int64,
+        np.int32,
+        np.int16,
+        np.uint8,
+    )
+
+    chosen = None
+    lhs_np = None
+    rhs_np = None
+
+    # Try to interpret raw bytes with different dtypes until we find a match
+    for dt in candidates:
+        try:
+            lhs_data = np.frombuffer(a, dtype=dt)
+            rhs_data = np.frombuffer(b, dtype=dt)
+        except (ValueError, TypeError):
+            continue
+        # We expect same element count on both sides and an even number (mins+maxs)
+        if lhs_data.size == rhs_data.size and lhs_data.size > 0 and (lhs_data.size % 2 == 0):
+            chosen = dt
+            lhs_np = lhs_data
+            rhs_np = rhs_data
+            break
+
+    if chosen is None:
+        raise RuntimeError(
+            "mpi_minmax: could not autodetect buffer dtype/size. "
+            f"bytes-left={len(a)}, bytes-right={len(b)}. "
+            "Ensure local_minmax returns same dtype/layout on all ranks."
+        )
```
Hi, thank you in advance for the PR again. I have one small suggestion: instead of trying to autodetect the data type and size with exceptions, I would recommend using closures to capture the buffer sizes and data types.

This should speed up the function and eliminate the need to capture exceptions. The operation will have to be freed after each use, but that is a common practice due to the limited number of self-defined operations available. 

You can check a similar example [in our implementation of _derived_op to handle custom MPI data types on reduction operations.](https://github.com/helmholtz-analytics/heat/blob/main/heat/core/communication.py#L805)

ivansherbakov9 commented:
@ClaudiaComito @JuanPedroGHM 
This commit implements the reviewers' suggestions:

- Moved the min/max MPI-op factory into the communications module.
- Implemented a closure-based `_minmax_op(dtype, count, offset)` that creates an MPI.Op specialized for the buffer dtype/count.
- ptp() now creates the MPI.Op on demand (and frees it after use) instead of relying on runtime autodetection within the callback.
- Added unit tests for the `_minmax_op` factory and made small doc updates.
- Ensured the reduction callback no longer raises exceptions at runtime and that dtypes/sizes are determined before registering the operation.

All local tests and pre-commit hooks ran successfully.

JuanPedroGHM commented on heat/core/statistics.py:
```diff
@@ -1739,6 +1740,187 @@ def _create_sketch(
+    >>> ht.ptp(a, axis=0)
+    DNDarray([9., 9., 9.], dtype=ht.float32, device=cpu:0, split=None)
+    >>> ht.ptp(a, axis=1)
+    DNDarray([2., 2., 2., 2.], dtype=ht.float32, device=cpu:0, split=None)
```
An example of how the axis argument works with a tuple would be great.

JuanPedroGHM commented on heat/core/tests/test_statistics.py:
```diff
@@ -1327,6 +1327,75 @@ def test_percentile_sketched(self):
+        self.assertAlmostEqual(ht.ptp(v), np.ptp(v_np), places=5)
+
+        v = ht.resplit(v, 0)
+        self.assertAlmostEqual(ht.ptp(v), np.ptp(v_np), places=5)
```
You can probably change all the ```assertAlmostEqual``` operations to ```assertEqual```. No floating point arithmetic happens.

JuanPedroGHM reviewed (commented):
Changes look good, there are just minor comments. I'll run the CI to see if something else comes up.
